File: python/say/say.py
def say(number):
    CHUNK_NAMES = ['', ' thousand', ' million', ' billion']

    number = int(number)

    if number < 0 or number > 999999999999:
        raise ValueError('Invalid number!')

    if number == 0:
        return 'zero'

    chunks = [int(str(number)[::-1][i:i + 3][::-1])
              for i in range(0, len(str(number)), 3)]
    parts = list(map(lambda e: (say_within_thousand(
        e[0]) + e[1]), filter(lambda e: e[0] != 0, zip(chunks, CHUNK_NAMES))))[::-1]
    if 0 < number % 1000 < 100 and number >= 1000:
        parts.insert(-1, 'and')
    return ' '.join(parts)


def say_within_thousand(n):
    ONE_NAMES = [None, 'one', 'two', 'three', 'four',
                 'five', 'six', 'seven', 'eight', 'nine']
    TEN_NAMES = [None, 'ten', 'twenty', 'thirty', 'forty',
                 'fifty', 'sixty', 'seventy', 'eighty', 'ninety']
    TEEN_NAMES = ['eleven', 'twelve', 'thirteen', 'fourteen',
                  'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen']

    parts = []

    hundred = n // 100
    if hundred > 0:
        parts.append(ONE_NAMES[hundred] + ' hundred')

    n %= 100
    if n > 0:
        if parts:
            parts.append('and')

        if 11 <= n <= 19:
            parts.append(TEEN_NAMES[n - 11])
        else:
            part = ''

            ten = n // 10
            if ten > 0:
                part += TEN_NAMES[ten]

            one = n % 10
            if one > 0:
                if part:
                    part += '-'
                part += ONE_NAMES[one]

            if part:
                parts.append(part)

    return ' '.join(parts)

File: java/say/src/main/java/Say.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Say {
  static final String[] CHUNK_NAMES = {
    "", "thousand", "million", "billion", "trillion", "quadrillion", "quintillion"
  };
  static final String[] ONE_NAMES = {
    "", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
  };
  static final String[] TEN_NAMES = {
    "", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
  };
  static final String[] TEEN_NAMES = {
    "eleven",
    "twelve",
    "thirteen",
    "fourteen",
    "fifteen",
    "sixteen",
    "seventeen",
    "eighteen",
    "nineteen"
  };

  public String say(long number) {
    if (number < 0 || number > 999_999_999_999L) {
      throw new IllegalArgumentException();
    }

    if (number == 0) {
      return "zero";
    }

    List<Integer> values = new ArrayList<>();
    while (number != 0) {
      values.add((int) (number % 1000));
      number /= 1000;
    }

    String[] parts =
        IntStream.range(0, values.size())
            .filter(i -> values.get(i) != 0)
            .mapToObj(
                i ->
                    sayWithinThousand(values.get(i))
                        + (CHUNK_NAMES[i].isEmpty() ? "" : " ")
                        + CHUNK_NAMES[i])
            .toArray(String[]::new);

    return IntStream.range(0, parts.length)
        .mapToObj(i -> parts[parts.length - 1 - i])
        .collect(Collectors.joining(" "));
  }

  String sayWithinThousand(int n) {
    List<String> parts = new ArrayList<>();

    if (n >= 100) {
      parts.add(ONE_NAMES[n / 100] + " hundred");
    }

    int rest = n % 100;
    if (rest != 0) {
      if (rest >= 11 && rest <= 19) {
        parts.add(TEEN_NAMES[rest - 11]);
      } else {
        StringBuilder part = new StringBuilder();

        int ten = rest / 10;
        if (ten != 0) {
          part.append(TEN_NAMES[ten]);
        }

        int one = rest % 10;
        if (one != 0) {
          if (!part.isEmpty()) {
            part.append("-");
          }
          part.append(ONE_NAMES[one]);
        }

        if (!part.isEmpty()) {
          parts.add(part.toString());
        }
      }
    }

    return String.join(" ", parts);
  }
}
